src/memray/_memray/snapshot.cpp

```diff
@@ -407,6 +420,163 @@ HighWaterMarkAggregator::visitAllocations(const allocation_callback_t& callback)
+        size_t generation_allocated)
+{
+    // We hold one reference, and the IntervalTree may or may not hold others.
+    // We use a count of 0 for all but the last deallocation of a range so that
+    // partial deallocations won't affect the count of allocations by location.
+    bool fully_deallocated = allocation_ptr.use_count() == 1;
+    recordDeallocation(
+            extractKey(*allocation_ptr),
+            (fully_deallocated ? 1 : 0),
+            bytes_deallocated,
+            generation_allocated);
+}
+
+void
+AllocationLifetimeAggregator::recordDeallocation(
+        const HighWaterMarkLocationKey& key,
+        size_t count_delta,
+        size_t bytes_delta,
+        size_t generation)
+{
+    if (d_num_snapshots == generation) {
+        // Allocated and deallocated within the same snapshot. We can ignore this.
+        return;
+    }
+
+    auto& counts = d_allocation_history[std::make_tuple(generation, d_num_snapshots, key)];
+    counts.first += count_delta;
+    counts.second += bytes_delta;
+}
+
+void
+AllocationLifetimeAggregator::captureSnapshot()
+{
+    ++d_num_snapshots;
+}
+
+std::vector<AllocationLifetime>
+AllocationLifetimeAggregator::generateIndex()
+{
+    struct KeyHash
+    {
+        size_t operator()(const std::pair<size_t, HighWaterMarkLocationKey>& key) const
+        {
+            size_t ret = HighWaterMarkLocationKeyHash{}(std::get<1>(key));
+            ret = (ret << 1) ^ std::get<0>(key);
+            return ret;
+        }
+    };
+
+    std::unordered_map<std::pair<size_t, HighWaterMarkLocationKey>, std::pair<size_t, size_t>, KeyHash>
+            leaks;
+
+    // Leaked simple allocations
+    for (const auto& [ptr, allocation_and_generation] : d_ptr_to_allocation) {
+        (void)ptr;
+        const auto& [allocation, generation] = allocation_and_generation;
+        auto& counts = leaks[std::make_pair(generation, extractKey(allocation))];
+        counts.first += 1;
+        counts.second += allocation.size;
+    }
+
+    // Leaked range allocations
+    std::unordered_set<void*> leaked_mappings;
+    for (const auto& [interval, allocation_ptr_and_generation] : d_mmap_intervals) {
+        const auto& [allocation_ptr, generation] = allocation_ptr_and_generation;
+        auto& counts = leaks[std::make_pair(generation, extractKey(*allocation_ptr))];
+
+        // Ensure we only count each allocation once, even if it's been split.
+        auto inserted = leaked_mappings.insert(allocation_ptr.get()).second;
+        counts.first += (inserted ? 1 : 0);
+        counts.second += interval.size();
+    }
+
+    std::vector<AllocationLifetime> ret;
+
+    // Things that were leaked
+    for (const auto& [when_where, how_much] : leaks) {
+        const auto& [allocated_before, key] = when_where;
+        const auto& [n_allocations, n_bytes] = how_much;
+        ret.push_back({allocated_before, static_cast<size_t>(-1), key, n_allocations, n_bytes});
+    }
+
+    // Things that weren't leaked
```

Comment:
I think we can discuss this offline, but I wonder if we can somehow handle true leaks in a closer fashion as the rest by just inserting them in `d_allocation_history` with `SIZE_MAX` or something in the `deallocatedBeforeSnapshot` field. Not that it really changes things a lot, but when reviewing this I kept wondering what makes true leaks so special with respect of the data model we are constructing here

Reply:
Discussed offline: we can't do that because the interface doesn't expose a way to signal "I'm done adding allocations". We could implicitly make the first call to `generateIndex` also mark that no more allocations can be added and merge all of the leaks into `d_allocation_history`, but that feels very weird. After our offline discussion, we decided to keep `generateIndex` side effect free.